### PR TITLE
PreconditionRelaxation: vmult/step w. relaxation

### PR DESCRIPTION
**Pull request by peterrum**
Add specialization for preconditioners with:
```
void vmult(dst, src, relax);
void step(dst, src, relax);
```

### CODE REVIEW

File: include/deal.II/lac/precondition.h [hunk 1]
```diff
@@ -569,6 +569,28 @@ namespace internal
...
+      template <typename U>
+      static decltype(
+        std::declval<U const>().step(std::declval<VectorType &>(),
+                                     std::declval<const VectorType &>(),
+                                     std::declval<const double>()))
+      detect(const U &);
+
+    public:
+      static const bool value =
+        !std::is_same<bool, decltype(detect(std::declval<T>()))>::value;
```
Comment (bangerth) [inline include/deal.II/lac/precondition.h:588]: Let's turn this around to make it read like "if this complicated type is 'bool'":
```suggestion
        !std::is_same<decltype(detect(std::declval<T>())),bool>::value;
```

**Review by bangerth — COMMENTED** [review]
Looks good. It's time we get to use C++20 concepts :-)

I assume you're going to test this in a follow-up patch somewhere?

**Comment by masterleinad** [comment]
> Looks good. It's time we get to use C++20 concepts :-)

We could just roll our own `is_detected` for now.

**Comment by bangerth** [comment]
I was thinking of `requires` clauses and proper concepts.

How does `is_detected` work?

**Comment by peterrum** [comment]
@bangerth I used the following code as template:

https://github.com/dealii/dealii/blob/ad71f7912936897a7d9409bd910eae1cdff9504c/include/deal.II/matrix_free/type_traits.h#L215-L241

I don't know all details SFINAE. I don't know why these lines were written like this. Maybe we could proceed with this PR like it is and postpone a possible clean up to later: there are some function in `matrix_free/type_traits.h` and in `lac/precondition.h` like this.

**Comment by peterrum** [comment]
> I assume you're going to test this in a follow-up patch somewhere?

Yes.

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
Yes, if there is existing art that does it the same way, let's stick with it!

**Comment by masterleinad** [comment]
> I was thinking of `requires` clauses and proper concepts.
> 
> How does `is_detected` work?

It pretty much does exactly what we are doing here but standardized (and shorter), see https://en.cppreference.com/w/cpp/experimental/is_detected. It might never hit the standard since there are concepts in `C++20`. Since it will take a good amount of time before requiring C++20 using a hand-rolled `is_detected` might be a good cleanup in the meantime.

**Comment by bangerth** [comment]
I would accept a patch based on this.

**Pull request merged**